api/v1alpha1/volumereplicationgroup_types.go

```diff
@@ -18,53 +18,32 @@ package v1alpha1
 
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	vrv1alpha1 "github.com/kube-storage/volume-replication-operator/api/v1alpha1"
 )
 
-// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
-// VolumeTakeoverControlType -- takeover control when desiredCluster is different from affinedCluster
-type VolumeTakeoverControlType string
-
-// VolumeTakeoverControlType definitions
-const (
-	// Force promote the volume in a WAN DR setting
-	ForcePromote VolumeTakeoverControlType = "ForcePromote"
-)
-
-// VolumeReplicationGroupSpec defines the desired state of VolumeReplicationGroup
+// VolumeReplicationGroup (VRG) spec declares the desired replication class
+// and replication state of all the PVCs owned by a given application, which
+// are identified via the given application label selector.  The VRG operator
+// creates children VolumeReplication (VR) CRs for each PVC of the application,
+// with the desired replication class and replication image state (primary or
+// secondary).
 type VolumeReplicationGroupSpec struct {
-	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 
 	// Name of the application to replicate
-	ApplicationName string `json:"applicationName"`
-
-	// Application label that is used to identify its PVs
-	ApplicationLabels metav1.LabelSelector `json:"applicationLabels"`
-
-	// Cluster ID in ClusterPeers that has best storage performance affinity for the application
-	AffinedCluster string `json:"affinedCluster"`
-
-	// Desired cluster that should takeover the application from current active cluster.
-	// May be set either to:
-	// - secondary cluster ID (to migrate application or takeover application in case of a disaster)
-	// - nil (if application is not active in the affined cluster, takeback to affined cluster)
-	// +optional
-	DesiredCluster string `json:"desiredCluster,omitempty"`
+	ApplicationName string `json:"applicationName,omitempty"`
 
-	// Volume Takeover Control: ForcePromote
-	// +optional
-	VolumeTakeoverControl VolumeTakeoverControlType `json:"volumeTakeoverControl,omitempty"`
+	// Application label selector that is used to identify all its PVCs
+	ApplicationLabelSelector metav1.LabelSelector `json:"applicationLabelSelector"`
 
-	// List of ClusterPeers
-	// For Metro DR only: a single ClusterPeers
-	// For WAN DR only: one or more ClusterPeers
-	// For MetroDR and WAN DR: one Metro DR ClusterPeer and one or more WAN DR ClusterPeers.
-	ClusterPeersList []string `json:"clusterPeersList"`
+	// ReplicationClass of all volumes in this replication group
+	VolumeReplicationClass string `json:"volumeReplicationClass"`
```

Comment:
(discussion) An VRG, based on the selector can end up with different PVCs from different providers. Thus causing the ReplicationClass to be an array instead. Matching the replicationClass to a PVC using the provider name where possible internally.

Another example could be different PVCs needing different ReplicationClass, due to differing replication policy (schedule for now), but may belong to the same selector/application.

I am unsure how we handle the use-cases in the future, for now 1:1 is good, hence opening this as a discussion for the time being.

Reply:
Good point.  This leads me to question whether the ReplicationClass spec should really be part of the PVC spec (as opposed to the VR spec), which then leads to the FAQ: can the features of ReplicationClass be implemented via StorageClass?

Reply:
> Good point. This leads me to question whether the ReplicationClass spec should really be part of the PVC spec (as opposed to the VR spec), which then leads to the FAQ: can the features of ReplicationClass be implemented via StorageClass?

(I am going to assume "ReplicationClass spec should really be part of the PVC Spec.." as "VolumeReplication spec should...")

The case is as follows,
- We need(ed) a VR CRD to aid addition of certain user desired/controlled properties to a replicated PVC
- These properties are not globally true for all PVCs (hence making them optional in the PVC spec is a no-go)
- (also, changing the PVC spec is a more time consuming and non-backward compatible)
- Once we needed a VR, adding the replication parameters for the CSI driver in question does not make sense in the StorageClass, as these are used exclusively by VR and rest of the parameters by the PVC, during creation
- So we end up with a replication class that holds these separate from the StorageClass

Should help the FAQ in the future 🤞🏾